### PR TITLE
Implement E741

### PR DESCRIPTION
**Pull request by harupy**
Test:

```bash
> flake8 --version
5.0.4 (mccabe: 0.7.0, pycodestyle: 2.9.1, pyflakes: 2.5.0) CPython 3.8.5 on Linux

> flake8 --select E741 resources/test/fixtures/E741.py
resources/test/fixtures/E741.py:3:1: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:4:1: E741 ambiguous variable name 'I'
resources/test/fixtures/E741.py:5:1: E741 ambiguous variable name 'O'
resources/test/fixtures/E741.py:8:4: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:10:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:11:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:16:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:25:12: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:26:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:30:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:33:18: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:34:9: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:40:8: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:49:16: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:63:22: E741 ambiguous variable name 'l'

> cargo run -- --select=E741 resources/test/fixtures/E741.py
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s
     Running `target/debug/ruff --select=E741 resources/test/fixtures/E741.py`
resources/test/fixtures/E741.py:3:1: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:4:1: E741 ambiguous variable name 'I'
resources/test/fixtures/E741.py:5:1: E741 ambiguous variable name 'O'
resources/test/fixtures/E741.py:6:1: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:8:4: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:9:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:10:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:11:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:16:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:20:8: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:25:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:26:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:30:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:33:9: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:34:9: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:40:8: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:49:16: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:58:20: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:63:1: E741 ambiguous variable name 'l'
```

### CODE REVIEW

File: resources/test/fixtures/E741.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+l = 0
+O: int = 0
+
+a, O = 0, 1
+[a, O] = 0, 1
+a, *O = 0, 1, 2
+a = O = 0
+
+o = 0
+i = 0
```
Comment (harupy) [inline resources/test/fixtures/E741.py:12]: Looks like the following cases are not covered:

```
for I in range(3):
    pass

with ctx() as I:
  pass

nonlocal I
global I
```

Comment (harupy) [reply resources/test/fixtures/E741.py:12]: https://github.com/PyCQA/pycodestyle/blob/10a4427c75740717b43448339fcf71f11bc33d1a/pycodestyle.py#L1482

```
@register_check
def ambiguous_identifier(logical_line, tokens):
    r"""Never use the characters 'l', 'O', or 'I' as variable names.

    In some fonts, these characters are indistinguishable from the
    numerals one and zero. When tempted to use 'l', use 'L' instead.

    Okay: L = 0
    Okay: o = 123
    Okay: i = 42
    E741: l = 0
    E741: O = 123
    E741: I = 42

    Variables can be bound in several other contexts, including class
    and function definitions, 'global' and 'nonlocal' statements,
    exception handlers, and 'with' and 'for' statements.
    In addition, we have a special handling for function parameters.

    Okay: except AttributeError as o:
    Okay: with lock as L:
    Okay: foo(l=12)
    Okay: for a in foo(l=12):
    E741: except AttributeError as O:
    E741: with lock as l:
    E741: global I
    E741: nonlocal l
    E741: def foo(l):
    E741: def foo(l=12):
    E741: l = foo(l=12)
    E741: for l in range(10):
    E742: class I(object):
    E743: def l(x):
    """
```

File: src/message.rs [hunk 2]
```diff
@@ -20,7 +20,7 @@ impl Ord for Message {
         (&self.filename, self.location.row(), self.location.column()).cmp(&(
             &other.filename,
             other.location.row(),
-            self.location.column(),
+            other.location.column(),
```
Comment (charliermarsh) [inline src/message.rs:23]: Oof, thanks!

File: src/check_ast.rs [hunk 3]
```diff
@@ -372,10 +385,30 @@ where
...
+                        .extend(checks::check_ambiguous_variable_name(target));
+                }
+            }
+            StmtKind::With { items, .. } | StmtKind::AsyncWith { items, .. } => {
+                if self.settings.select.contains(&CheckCode::E741) {
+                    for item in items {
+                        if let Some(vars) = &item.optional_vars {
+                            self.checks
+                                .extend(checks::check_ambiguous_variable_name(vars))
+                        }
```
Comment (charliermarsh) [inline src/check_ast.rs:408]: Do you have a sense for how many of these would be caught by instead hooking into `handle_node_store`? Could it help reduce the number of cases we have to handle?

File: src/message.rs (hunk 2 above)
Comment (harupy) [reply src/message.rs:23]: @charliermarsh I opened https://github.com/charliermarsh/ruff/pull/152 just to fix this line.

File: src/check_ast.rs (hunk 3 above)
Comment (harupy) [reply src/check_ast.rs:408]: > Do you have a sense for how many of these would be caught by instead hooking into handle_node_store?

@charliermarsh Thanks for the comment. By `hooking into handle_node_store`, you meant something like below? 

```rust
    fn handle_node_store(&mut self, expr: &Expr, parent: Option<&Stmt>) {
        if let ExprKind::Name { id, .. } = &expr.node {
            ...

            // For example, `should_check_ambiguous_variable_name` should be True
            // for Name in Assign.targets
            let should_check_ambiguous_variable_name = ...;
            if self.settings.select.contains(&CheckCode::E741)
                && should_check_ambiguous_variable_name
                && checks::is_ambiguous_name(id)
            {
                self.checks.push(Check::new(
                    CheckKind::AmbiguousVariableName(id.to_string()),
                    expr.location,
                ));
            }
```

**Comment by harupy** [comment]
```
> cargo run --select=E741, resources/test/fixtures/E741.py
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s
     Running `target/debug/ruff resources/test/fixtures/E741.py`
resources/test/fixtures/E741.py:3:1: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:4:1: E741 ambiguous variable name 'I'
resources/test/fixtures/E741.py:5:1: E741 ambiguous variable name 'O'
resources/test/fixtures/E741.py:6:1: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:8:4: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:9:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:10:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:11:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:16:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:20:8: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:25:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:26:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:30:5: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:33:9: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:34:9: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:40:8: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:40:14: E741 ambiguous variable name 'I'
resources/test/fixtures/E741.py:44:8: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:44:16: E741 ambiguous variable name 'I'
resources/test/fixtures/E741.py:48:9: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:48:14: E741 ambiguous variable name 'I'
resources/test/fixtures/E741.py:57:16: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:66:20: E741 ambiguous variable name 'l'
resources/test/fixtures/E741.py:71:1: E741 ambiguous variable name 'l'  <---
resources/test/fixtures/E741.py:71:1: E741 ambiguous variable name 'l'  <---
```

It looks like `handle_node_store` is called twice on the same except handler alias:

```python
# resources/test/fixtures/E741.py

...

try:
    pass
except ValueError as l:
    #                ^ this
    pass
```

File: src/check_ast.rs (hunk 3 above)
Comment (charliermarsh) [reply src/check_ast.rs]: Yeah that's roughly what I was suggesting!

**Comment by charliermarsh** [comment]
Hmm yeah, it is... if you look at `visit_excepthandler`, we do call it twice. That was based on the PyFlakes logic [here](https://github.com/PyCQA/pyflakes/blob/master/pyflakes/checker.py#L2227). Perhaps we can just get rid of the second `handle_node_store`. The current version, at least, is buggy in that `let definition = scope.values.get(name).cloned();` should be removing the name from the scope.

**Comment by charliermarsh** [comment]
What if we instead put the lint rule logic in the `ExprContext::Store` block here?

```
ExprKind::Name { ctx, .. } => match ctx {
    ExprContext::Load => self.handle_node_load(expr),
    ExprContext::Store => {
        let parent =
            self.parents[*(self.parent_stack.last().expect("No parent found."))];
        self.handle_node_store(expr, Some(parent));
    }
    ExprContext::Del => self.handle_node_delete(expr),
},
```

...instead of in the `handle_node_store` function.

**Review by charliermarsh — APPROVED** [review]
Great, thank you!

**Pull request merged**